CARDS-1954: The questionnaire editor fetches the questionnaire JSON too many times

veronikaslc opened the pull request:
To test[ CARDS-1954](https://phenotips.atlassian.net/browse/CARDS-1954):
- start in `prems` mode
- open network tab in browser
- open the CPESIC  questionnaire in the editor
- notice how many times `/Questionnaires/CPESIC.deep.json` is fetched (before was 10)

veronikaslc committed:
CARDS-1954: The questionnaire editor fetches the questionnaire JSON too many times

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -76,9 +78,16 @@ let Questionnaire = (props) => {
+  // When the questionnaire data is fetched, we inform the QuestionnaireContext
+  useEffect(() => {
+    if (data) {
+	  let vars = [];
```
Messy indentation

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -76,9 +78,16 @@ let Questionnaire = (props) => {
-  if (!data) {
-    fetchData();
-  }
+  // When the questionnaire data is fetched, we inform the QuestionnaireContext
+  useEffect(() => {
+    if (data) {
+	  let vars = [];
+      findQuestions(data, vars);
+      changeQuestionnaireContext(vars);
+    } else {
+      fetchData();
+    }
+  }, [data]);
```
I think two separate effects make more sense here:

```suggestion
  // First, fetch the questionnaire data
  useEffect(() => {
    fetchData();
  }, []);

  // When the questionnaire data is fetched, we inform the QuestionnaireContext
  useEffect(() => {
    if (data) {
      let vars = [];
      findQuestions(data, vars);
      changeQuestionnaireContext(vars);
    }
  }, [data]);
```

veronikaslc committed:
CARDS-1954: The questionnaire editor fetches the questionnaire JSON too many times

Better effect arrangements

Co-authored-by: Marta Girdea <marta@phenotips.org>

sashaandjic reviewed (approved):
tested
![image](https://user-images.githubusercontent.com/23243244/204430564-77709a84-0847-49c8-bff1-00920a7d1625.png)

marta- commented:
Additional *regression testing*:
- [ ] using the questionnaire designer, add and/or edit conditions
- [ ] create a form for that questionnaire and test that conditions modified in the designer work as expected

sashaandjic reviewed (approved):
tested:

BEFORE:
![image](https://user-images.githubusercontent.com/23243244/205980451-df7fa0f8-43f8-4271-8587-d0aae7fdaa0a.png)


NOW:
![image](https://user-images.githubusercontent.com/23243244/205980525-979ed54b-75a0-4358-b856-cfecfc7744ad.png)

sashaandjic commented:
Tested
> Additional **regression testing** needed:
> 
> * [-]  using the questionnaire designer, add and/or edit conditions
> * [-]  create a form for that questionnaire and test that conditions modified in the designer work as expected

sdumitriu reviewed (changes requested):
This fails to keep the context updated as questions are added/modified/deleted.

- create a new questionnaire
- add a question "q1" with the text "old q1"
- add a section, add a condition for the section -> q1 does not show up as an available question for the condition
- reload the editor
- edit the condition -> "old q1" shows up, but don't select it, cancel the edit dialog
- edit q1 to change the text to "new q1"
- edit the condition -> still "old q1" shows up

veronikaslc committed:
CARDS-1954: The questionnaire editor fetches the questionnaire JSON too many times
Fixed context change for question create/update/delete

marta- reviewed (changes requested):
**Regression**:
- Create a new questionnaire
- Add 2 questions: q1 and q2
- Add a section
- Start adding a conditional to the section:
  - for *Operand a* choose q1
  - for *Operand b* switch the toggle from "Value" to "Question id" and click on the dropdown to select q2 -> **the dropdown is empty**

![image](https://user-images.githubusercontent.com/651980/207703465-344470cf-c9b0-49f5-8a3f-69cb28963770.png)

marta- commented:
**Regression 2:**
- Create a new questionnaire
- Add 2 questions: q1 and q2
- Add a section
- Add q3 to the section
- After adding q3 to the section, start adding a conditional to the section:
  - click on the *Operand a* dropdown to reveal available questions - **only q3 is displayed** (due to the questionnaire context being populated with **only** the section's contents)
- Reload the page and re-attempt adding a conditional: all 3 questions (q1, q2, q3) are available for selection for *Operand a* - correct

veronikaslc committed:
CARDS-1954: The questionnaire editor fetches the questionnaire JSON too many times
Fixed regression on conditional second operand for question ID selector dropdown

veronikaslc commented:
fixed first Regression

veronikaslc committed:
CARDS-1954: The questionnaire editor fetches the questionnaire JSON too many times
Fixed context update on question deletion

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -113,6 +129,16 @@ let Questionnaire = (props) => {
     });
   }, []);
 
+  let findQuestions = (json, result) =>  {
+    Object.entries(json || {}).forEach(([k,e]) => {
+      if (e?.['jcr:primaryType'] == "cards:Question") {
+        result.push({name: e['@name'], text: e['text']});
+      } else if (typeof(e) == 'object') {
+        findQuestions(e, result);
+      }
+    })
+  }
+
```
This function is repeated 3 times. It should be defined only once and reused as necessary.

veronikaslc replied:
Only 2 times, I will move it up out of the components scopes.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/Questionnaires.jsx:
```diff
@@ -23,12 +23,15 @@ import NewQuestionnaireDialog from "../questionnaireEditor/NewQuestionnaireDialo
 
   if (entry) {
-    return <Questionnaire id={entry[1]} key={location.pathname} contentOffset={props.contentOffset}/>;
+    return <QuestionnaireProvider>
```
Why is the `<QuestionnaireProvider>` all the way here instead of simply wrapping `<QuestionnaireContents>` in Questionnaire.jsx?

veronikaslc replied:
Because `useQuestionnaireWriterContext` must be used within a `QuestionnaireProvider`.

marta- replied:
Refactored in 069afae

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -133,7 +159,7 @@ let Questionnaire = (props) => {
           entryType="Questionnaire"
           variant="icon"
-          onComplete={() => history.replace(baseUrl)}
+          onComplete={ () => { updateContext(); history.replace(baseUrl);} }
```
`updateContext()` is not necessary here, We're deleting the whole questionnaire an moving away from the page, so the context will no longer be in use.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/QuestionnaireItemCard.jsx:
```diff
@@ -139,6 +140,31 @@ let QuestionnaireItemCard = (props) => {
 
   const styles = useStyles();
 
+  let changeQuestionnaireContext = useQuestionnaireWriterContext();
+
+  let updateContext = (data) => {
+    let questionNames = [];
+    findQuestionsNames({data: data}, questionNames);
+    changeQuestionnaireContext((oldContext) => {
+        let newContext = oldContext;
+        questionNames.map(name => {
+            const index = newContext.findIndex(x => x.name == name);
+            newContext.splice(index, 1);
+        });
+        return newContext;
+    });
+  }
+
+  let findQuestionsNames = (json, result) =>  {
+    Object.entries(json || {}).forEach(([k,e]) => {
+      if (e?.['jcr:primaryType'] == "cards:Question") {
+        result.push(e['@name']);
+      } else if (typeof(e) == 'object') {
+        findQuestionsNames(e, result);
+      }
+    })
+  }
+
```
In addition to repeating `findQuestionsNames` a third time here, the whole functionality doesn't belong. As suggested by its name, the purpose of the `QuestionnaireItemCard` component is limited to displaying the box. Any context updates should be passed from the parent component via `onActionDone`, rather than being implemented here.

veronikaslc replied:
Is it not repeating, though it looks similar. Here it collects only names of questions to be deleted (line 161), on other places it fills up JSON object with names and labeles.

The decision of placing the context update functionality here came from the goal of having as little refactoring as possible.  Passing from the parent component via `onActionDone` would imply refactoring as `onActionDone` is in fact `reloadData` in `QuestionnaireEntry` parent component.  `reloadData` was designed for the purpose of updates only, not including delete, recursively calling itseld and `onActionDone` in the body. We would need to introduce the action specification, pass on the deleted data to update the context accordingly, add a new logical paths to deal with the cases of deletion, look into affects of parent `onActionDone`. If we do that we need to decide how we refactor `reloadData`, what is the role of `newData` parameter. Do we ever use it and when / where it was intended to be passed? We do not seem to pass `newData` explicitly and the history of introducing the related code goes back to 2 years ago[ commit](https://github.com/data-team-uhn/cards/commit/d0e62284ca28507fb3009100912cce46e0b1d5b0) devoted to the hightlighting and scrolling control. 

In short - personaly i'd prefer to keep it all in the paternt `QuestionnaireEntry` but there is alot of internal controls involed and refactoring/testing of it would imply testing all involved functionality.

marta- replied:
A simple rename of `reloadData` and adding the removal of the deleted question from the context on the `catch` branch should suffice.

marta- replied:
Refactored in 5954583

veronikaslc committed:
CARDS-1954: The questionnaire editor fetches the questionnaire JSON too many times
Reverted context update on the questionnaire deletion

veronikaslc committed:
CARDS-1954: The questionnaire editor fetches the questionnaire JSON too many times
Moved utility function out of the components scope

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireContext.jsx:
```diff
@@ -0,0 +1,71 @@
+ */
+export function QuestionnaireProvider(props) {
+  const [questions, setQuestions] = React.useState(DEFAULT_STATE);
+  const {additionalFormData, ...rest} = props
```
This looks like a leftover from copying and adapting FormContext, to be removed.

marta- replied:
Removed in 5d0f2c5

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/QuestionnaireItemCard.jsx:
```diff
@@ -139,6 +140,31 @@ let QuestionnaireItemCard = (props) => {
+        let newContext = oldContext;
+        questionNames.map(name => {
+            const index = newContext.findIndex(x => x.name == name);
+            newContext.splice(index, 1);
```
We're removing from the context the first question that matches the deleted one's name. But that might not actually be the deleted one. We can have questions with the same `@name`  in different sections, and if we delete the 2nd one from the questionnaire, the first question with that `@name` will be removed from context - **Bug**.

marta- replied:
Fixed in 5954583

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ConditionalValueInput.jsx:
```diff
@@ -83,8 +84,9 @@ let ConditionalValueInput = (props) => {
   let [ variables, setVariables ] = useState();
   let [ error, setError ] = useState();
 
+  let reader = useQuestionnaireReaderContext();
```
`reader` is not a suggestive name. I would use `questions` instead.

marta- replied:
Done in e6b2cfb

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -469,6 +497,14 @@ let QuestionnaireEntry = (props) => {
+  let updateContext = (data) => {
+    let vars = [];
+    findQuestions({data: data}, vars);
+    changeQuestionnaireContext((oldContext) => ([ ...oldContext, ...vars ]));
```
If we keep adding to the existing context, duplicates get created - **Bug.**

Example: Create a new questionnaire. Add questions `q1`, `q2`, add section `s`. Start adding a conditional in `s`. Note how the list of available questions is q1, q1, q2, q1, q2.

marta- replied:
Fixed in 5954583

marta- committed:
CARDS-1954: The questionnaire editor fetches the questionnaire JSON too many times
[minor] rename variable

veronikaslc commented:
Tested

veronikaslc commented:
Tested

marta- committed:
CARDS-1954: The questionnaire editor fetches the questionnaire JSON too many times
Remove unused prop

marta- committed:
CARDS-1954: The questionnaire editor fetches the questionnaire JSON too many times
Refactoring: Move the `<QuestionnaireProvider>` wrapper inside Questionnaire
Bug fix: Clear the context when leaving the page, to prevent questions from leaking into other questionnaire's context

marta- committed:
CARDS-1954: The questionnaire editor fetches the questionnaire JSON too many times
Bug fix: existing questions get added to the context again whenever a new questionnaire entry is added
Bug fix: when there are 2 questions with the same @name and the second one is deleted from the questionnaire, the first one is removed from the context
Refactoring: Move context update from QuestionnaireItemCard to QuestionnaireEntry

sashaandjic reviewed (approved):
tested

sdumitriu reviewed (approved):
This works well, but there are a few other things that are broken with conditionals when multiple questions have the same name. Perhaps we should reconsider how conditionals work, instead of using just the question name, we should use the full path.
